### PR TITLE
Added console input History, and trap Error in Try-Catch

### PR DESCRIPTION
**Pull request by mingfudesign**
Added $history to track latest 99 valid inputs
Added commands for showing console inputs
Added pre-amble at Console start to list available commands
Try-Catch traps Errors to prevent Console from exiting automatically

### CODE REVIEW

File: classes/console.php [hunk 1]
```diff
@@ -2,12 +2,12 @@
...
- * @author     Fuel Development Team
- * @license    MIT License
+ * @package	   Fuel
+ * @version	   1.9-dev
+ * @author	   Fuel Development Team
+ * @license	   MIT License
  * @copyright  2010 - 2019 Fuel Development Team
- * @link       https://fuelphp.com
+ * @link	   https://fuelphp.com
  */
```
Comment (WanWizard) [inline classes/console.php]: Please undo the changes in the header (no tabs).

File: classes/console.php [hunk 2]
```diff
@@ -64,6 +64,29 @@ public static function help()
 
 	}
 
+	private const MAX_HISTORY = 99;
+	private $history = [];
```
Comment (WanWizard) [inline classes/console.php]: Class variables should be defined at the top, before the constructor, not in the middle of the code.
Also, never use "private", use "protected". All Fuel classes should be fully extendable.

File: classes/console.php [hunk 3]
```diff
@@ -64,6 +64,29 @@ public static function help()
...
+
+	private function push_history($line)
```
Comment (WanWizard) [inline classes/console.php]: protected instead of private.

File: classes/console.php [hunk 4]
```diff
@@ -64,6 +64,29 @@ public static function help()
...
+	{
+		// delimit each line, allowing for copy & paste of history back into Console
+		$line .= ';';
+
+		array_push($this->history, $line);
+		$this->history = array_slice($this->history, -self::MAX_HISTORY);
+	}
+
+	private function pop_history()
```
Comment (WanWizard) [inline classes/console.php]: protected instead of private.

File: classes/console.php [hunk 5]
```diff
@@ -64,6 +64,29 @@ public static function help()
...
+	{
+		array_pop($this->history);
+	}
+
+	private function show_history()
```
Comment (WanWizard) [inline classes/console.php]: protected instead of private.

File: classes/console.php [hunk 6]
```diff
@@ -75,6 +98,8 @@ private function main()
 			PHP_OS
 		));
 
+		\Cli::write(['', 'Commands', ':q | quit - exit the console', ':h | history - show transcript', '']);
```
Comment (WanWizard) [inline classes/console.php]: No short array notation please, Fuel still requires PHP 5.6 compatibility.

File: classes/console.php [hunk 7]
```diff
@@ -64,6 +64,29 @@ public static function help()
...
+		array_pop($this->history);
+	}
+
+	private function show_history()
+	{
+		\Cli::write($this->history);
+		\Cli::write('');
+	}
+
 	private function main()
```
Comment (WanWizard) [inline classes/console.php]: Please while you're at it, please change this to protected too...

File: classes/console.php [hunk 8]
```diff
@@ -88,13 +113,19 @@ private function main()
 				continue;
 			}
 
-			if ($__line == 'quit')
+			if ($__line == ':q' || $__line == 'quit')
```
Comment (WanWizard) [inline classes/console.php]: Please use "or" instead of "||" ( coding standards )

File: classes/console.php [hunk 9]
```diff
@@ -88,13 +113,19 @@ private function main()
...
 			{
 				break;
 			}
+			elseif ($__line == ':h' || $__line == 'history')
```
Comment (WanWizard) [inline classes/console.php]: Please use "or" instead of "||" ( coding standards )

File: classes/console.php [hunk 10]
```diff
@@ -113,8 +144,11 @@ private function main()
 			{
 				$ret = eval("unset(\$__line); $__line;");
 			}
-			catch(\Exception $e)
+			catch(\Exception|\Error $e)
```
Comment (WanWizard) [inline classes/console.php]: Catching multiple exceptions isnt supported in PHP < 7.1, so please split this.

**Review by WanWizard — CHANGES_REQUESTED** [review]
Thanks for your contribution, much appreciated!

Could you please address the comments?

**Comment by mingfudesign** [comment]
I think, with the latest changes, all of the comments have been addressed. Please check.

**Comment by WanWizard** [comment]
Looks fine, thanks !

**Pull request merged**